### ISSUE TITLE
Cannot find classes at runtime

### ISSUE THREAD

**Issue by ackeri** [oracle/graal#618]
I'm trying to serialize non-java objects in GraalVM, and am writing out the fully qualified class names in order to reconstruct on the other side. There are still some outstanding issues when serializing, but of focus here, when I try to look up one of the classes, the JVM doesn't seem to be aware of them. 

ClassNotFoundException: com.oracle.truffle.api.vm.PolyglotValue$Interop

Which is exactly what I got for the class name of array.impl in:
Value array = polyglot.eval("js", "[1,2,42,4]");

Are these classes loaded somewhere unusual?

**Comment by chumer** [comment]
Yes these classes are loaded in a separate class loader. Language implementations are isolated from the application classpath. We currently don't support Java serialization of guest values.

Would serializing to json and reconstructing from there be an option?

**Comment by ackeri** [comment]
Ideally the type/language information could be preserved along the way, so that I could get a guest language object when deserializing. If the representation (or some higher level interface) of guest values isn't available in the classpath, then we can't instantiate instances of guest values using reflections.

**Comment by pekd** [comment]
> Ideally the type/language information could be preserved along the way, so that I could get a guest language object when deserializing.

But how would that work if the `Value` supports `execute` or `newInstance`? And what would happen if the `Value` represents a native pointer? In those cases the `Value` depends on the guest language's state; in the case of a native pointer it even depends on the memory of the JVM process. Even if the classes were available on the classpath, you couldn't fully support serialization of all possible types of `Value` objects.

**Comment by ackeri** [comment]
Sure, you couldn't support all serialization use cases, but many, many cases can be covered. 

Regarding your objections:
-Code is data, so executable Values can be serialized. For typed languages (or typed subsets of languages) you can often even just encode the type name. For existing tech, I believe Java RMI allows transmitting bytecode for class definitions. 
-Native pointers usually represent some abstract data structure that is typed. LLIR even contains most of this typing information. It isn't guaranteed by many native languages, but native pointers almost always point to a contained, explorable memory space that could be replicated to have the same meaning.

I'm sure there are cases where it doesn't make sense to serialize something, but the vast majority of data is theoretically serialize-able.

**Comment by ackeri** [comment]
Serializing to json or other formats works fine. The problem is when deserializing data, there isn't a good way that I've found to instantiate the object that should hold said data. I was hoping that Java serialization could be somewhat easily extended to the guest values, but that doesn't look to be in the cards. newInstance method on value would work in theory, but I've only got it to work for proxy-d java (most of my testing has been with js). Even then it is not ideal, because we to keep track of the metaobjects of all the things we need to be able to deserialize.
